wellcomeml/io/epmc/client.py

```diff
@@ -0,0 +1,302 @@
+
+        while go_to_next_page:
+            self.logger.debug(f"Querying results from "
+                              f"{page*params['pageSize']} to "
+                              f"{(page+1)*params['pageSize']}")
+
+            response = session.get(
+                epmc_search_url,
+                params=params
+            )
+
+            response_json = response.json()
+
+            current_page_result = response_json["resultList"]["result"]
+
+            if not current_page_result or only_first:
+                go_to_next_page = False
+            else:
+                params['cursorMark'] = response_json.get("nextCursorMark")
+
+            final_results += current_page_result
+            page += 1
+
+        if final_results:
+            return final_results[0] if only_first else final_results
+        else:
+            self.logger.warning(
+                "epmc.license.get_pm_metadata: no-results query=%s", params
+            )
+            return None
+
+    def search(self, session, query, result_type='core', page_size=10,
+               only_first=True, **kwargs):
+        """ Fetches metadata from EPMC's REST API for a query. """
+        # Sample URL for searching by DOI:
+        #   https://www.ebi.ac.uk/europepmc/webservices/rest/search?query=doi:10.1111/hiv.12674
+
+        params = {
+            "query": query,
+            "format": "json",
+            "resultType": result_type,
+            "pageSize": page_size
+        }
+
+        return self._execute_query(session, params, only_first, **kwargs)
+
+    def search_by_doi(self, session, doi, **kwargs):
+        query = "doi:%s" % doi
+        return self.search(session, query, only_first=True, **kwargs)
+
+    def search_by_pmid(self, session, pmid, **kwargs):
+        query = "ext_id:%s" % pmid
+        return self.search(session, query, only_first=True, **kwargs)
+
+    def search_by_pmids(self, session, pmids, **kwargs):
+        query = "ext_id:%s" % '"' + '" OR ext_id:"'.join(pmids) + '"'
+        return self.search(session, query, only_first=False, **kwargs)
+
+    def search_by_pmcid(self, session, pmcid, **kwargs):
+        if not pmcid.startswith('PMC'):
+            raise ValueError("Invalid PMCID: %s" % pmcid)
+        query = "pmcid:%s" % pmcid
+        return self.search(session, query, only_first=True, **kwargs)
+
+    def search_by_pmcids(self, session, pmcids, **kwargs):
+        query = "pmcid:%s" % '"' + '" OR pmcid:"'.join(pmcids) + '"'
+        return self.search(session, query, only_first=False, **kwargs)
+
+    def search_by_dois(self, session, dois, **kwargs):
+        """
+        Args:
+            session: requests.Session
+            dois: list of DOI strings
+
+        Yields:
+            Dicts of EPMC responses
+        """
+        query = '"' + '" OR "'.join(dois) + '"'
+        return self.search(session=session, query=query, only_first=False, **kwargs)
+
+    def get_full_text(self, session, pmid):
```

Comment:
This function is not tested - would be good to add a test, as I suspect it's a tricky one.

Reply:
Same as the others, just need to ensure search is called with the right args. This assumes search is tested. Actually I noticed that the tests call the API which needs to change as well.

Reply:
I think we do want to test calling the API? As in, if they modify their response, we would like to know that it is broken. Probably mark the test as `pytest.mark.integrations`

Reply:
7c3b2293a6fa061eb878e5b78debd81b27b51a6f